ePetra: OMP statement incompatible with OMP 4.0 / GCC 9 data sharing rules

## Bug Report
@trilinos/ePetra

### Description
https://www.gnu.org/software/gcc/gcc-9/porting_to.html
> GCC releases before 9 were implementing an OpenMP 3.1 data sharing rule that const qualified variables without mutable member are predetermined shared, ... GCC 9 started implementing the OpenMP 4.0 and later behavior. When not using default clause or when using default(shared), this makes no difference, but if using default(none), previously the choice was not specify the const qualified variables on the construct at all, ...

### Steps to Reproduce
Try to build with GCC 9
```
[  299s] /home/abuild/rpmbuild/BUILD/Trilinos-trilinos-release-12-10-1/packages/epetra/src/Epetra_CrsGraph.cpp: In member function 'int Epetra_CrsGraph::OptimizeStorage()':
[  299s] /home/abuild/rpmbuild/BUILD/Trilinos-trilinos-release-12-10-1/packages/epetra/src/Epetra_CrsGraph.cpp:1949:7: error: 'numMyBlockRows' not specified in enclosing 'parallel'
[  299s]  1949 |       for(int i = 0; i < numMyBlockRows; i++) {
[  299s]       |       ^~~
[  299s] /home/abuild/rpmbuild/BUILD/Trilinos-trilinos-release-12-10-1/packages/epetra/src/Epetra_CrsGraph.cpp:1947:9: error: enclosing 'parallel'
[  299s]  1947 | #pragma omp parallel for default(none) shared(indexOffset,all_indices,indices)
[  299s]       |         ^~~
[  299s] /home/abuild/rpmbuild/BUILD/Trilinos-trilinos-release-12-10-1/packages/epetra/src/Epetra_CrsGraph.cpp:1989:6: error: 'numMyBlockRows' not specified in enclosing 'parallel'
```

--- a/packages/epetra/src/Epetra_CrsGraph.cpp
+++ b/packages/epetra/src/Epetra_CrsGraph.cpp
@@ -1938,7 +1938,7 @@ int Epetra_CrsGraph::OptimizeStorage() {
 
     if (!(StaticProfile())) {
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(indexOffset,all_indices,indices)
+#pragma omp parallel for default(none) shared(indexOffset,all_indices,indices,numMyBlockRows)
 #endif
       for(int i = 0; i < numMyBlockRows; i++) {
   int numColIndices = indexOffset[i+1] - indexOffset[i];
@@ -1978,7 +1978,7 @@ int Epetra_CrsGraph::OptimizeStorage() {
       int * indexOffset = CrsGraphData_->IndexOffset_.Values();
 
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(indexOffset,old_all_indices,new_all_indices)
+#pragma omp parallel for default(none) shared(indexOffset,old_all_indices,new_all_indices,numMyBlockRows)
 #endif
      for(int i = 0; i < numMyBlockRows; i++) {
        int numColIndices = indexOffset[i+1] - indexOffset[i];

--- a/packages/epetra/src/Epetra_CrsGraph.cpp
+++ b/packages/epetra/src/Epetra_CrsGraph.cpp
@@ -1875,7 +1875,7 @@ int Epetra_CrsGraph::MakeIndicesLocal(const Epetra_BlockMap& domainMap, const Ep
 //==============================================================================
 int Epetra_CrsGraph::OptimizeStorage() {
   int NumIndices;
-  const int numMyBlockRows = NumMyBlockRows();
+  int numMyBlockRows = NumMyBlockRows();
 
   Epetra_CrsGraphData::IndexData<int>& Data = CrsGraphData_->Data<int>();
 

--- a/packages/epetra/src/Epetra_CrsMatrix.cpp
+++ b/packages/epetra/src/Epetra_CrsMatrix.cpp
@@ -1316,8 +1316,8 @@ int Epetra_CrsMatrix::OptimizeStorage() {
       All_Values_ = new double[numMyNonzeros];
       if(All_Values_ == 0) throw ReportError("Error with All_Values_ allocation.", -99);
 
-      const int *const  IndexOffset = Graph().IndexOffset();
-      const int numMyRows = NumMyRows_;
+      const int* IndexOffset = Graph().IndexOffset();
+      int numMyRows = NumMyRows_;
       double ** Values_s = Values_;
       double * All_Values_s = All_Values_;
 #ifdef EPETRA_HAVE_OMP
@@ -3328,7 +3328,7 @@ if (StorageOptimized() && Graph().StorageOptimized()) {
     char matdescra[6] = "G//C/";
     mkl_dcsrmv(&transa, &m, &NumCols, &alpha, matdescra, values, Indices, IndexOffset, IndexOffset + 1, x, &beta, y);
 #elif defined(EPETRA_HAVE_OMP)
-  const int numMyRows = NumMyRows_;
+  int numMyRows = NumMyRows_;
 #pragma omp parallel for default(none) shared(IndexOffset,values,Indices,y,x,numMyRows)
      for (int row=0; row<numMyRows; ++row)
         {
@@ -3371,7 +3371,7 @@ if (StorageOptimized() && Graph().StorageOptimized()) {
     int* NumEntriesPerRow = Graph().NumIndicesPerRow();
     int** Indices = Graph().Indices();
     double** srcValues = Values();
-        const int numMyRows = NumMyRows_;
+    int numMyRows = NumMyRows_;
 
     // Do actual computation
 #ifdef EPETRA_HAVE_OMP
@@ -3391,7 +3391,7 @@ if (StorageOptimized() && Graph().StorageOptimized()) {
   }
   else { // Case where StorageOptimized is incompatible:  Use general accessors.
 
-    const int numMyRows = NumMyRows_;
+    int numMyRows = NumMyRows_;
 
     // Do actual computation
 #ifdef EPETRA_HAVE_OMP
@@ -3513,9 +3513,9 @@ void Epetra_CrsMatrix::GeneralMM(double ** X, int LDX, double ** Y, int LDY, int
     return;
     }
 
-    double ** const xp = X;
-    double ** const yp = Y;
-    const int numMyRows = NumMyRows_;
+    double ** xp = X;
+    double ** yp = Y;
+    int numMyRows = NumMyRows_;
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(IndexOffset,Indices,values,NumVectors,numMyRows,xp,yp)
 #endif
@@ -3541,9 +3541,9 @@ void Epetra_CrsMatrix::GeneralMM(double ** X, int LDX, double ** Y, int LDY, int
     int* NumEntriesPerRow = Graph().NumIndicesPerRow();
     int** Indices = Graph().Indices();
     double** srcValues = Values();
-    double ** const xp = X;
-    double ** const yp = Y;
-    const int numMyRows = NumMyRows_;
+    double ** xp = X;
+    double ** yp = Y;
+    int numMyRows = NumMyRows_;
 
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(NumEntriesPerRow,Indices,srcValues,NumVectors,numMyRows,xp,yp)
@@ -3563,9 +3563,9 @@ void Epetra_CrsMatrix::GeneralMM(double ** X, int LDX, double ** Y, int LDY, int
   }
   else {
 
-    double ** const xp = X;
-    double ** const yp = Y;
-    const int numMyRows = NumMyRows_;
+    double ** xp = X;
+    double ** yp = Y;
+    int numMyRows = NumMyRows_;
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(NumVectors,numMyRows,xp,yp)
 #endif

--- a/packages/epetra/src/Epetra_CrsMatrix.cpp
+++ b/packages/epetra/src/Epetra_CrsMatrix.cpp
@@ -1321,7 +1321,7 @@ int Epetra_CrsMatrix::OptimizeStorage() {
       double ** Values_s = Values_;
       double * All_Values_s = All_Values_;
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(Values_s,All_Values_s)
+#pragma omp parallel for default(none) shared(Values_s,All_Values_s,numMyRows,IndexOffset)
 #endif
       for (int i=0; i<numMyRows; i++) {
         int NumEntries = Graph().NumMyIndices(i);
@@ -3329,7 +3329,7 @@ if (StorageOptimized() && Graph().StorageOptimized()) {
     mkl_dcsrmv(&transa, &m, &NumCols, &alpha, matdescra, values, Indices, IndexOffset, IndexOffset + 1, x, &beta, y);
 #elif defined(EPETRA_HAVE_OMP)
   const int numMyRows = NumMyRows_;
-#pragma omp parallel for default(none) shared(IndexOffset,values,Indices,y,x)
+#pragma omp parallel for default(none) shared(IndexOffset,values,Indices,y,x,numMyRows)
      for (int row=0; row<numMyRows; ++row)
         {
      const int curOffset = IndexOffset[row];
@@ -3375,7 +3375,7 @@ if (StorageOptimized() && Graph().StorageOptimized()) {
 
     // Do actual computation
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(NumEntriesPerRow,Indices,srcValues,y,x)
+#pragma omp parallel for default(none) shared(NumEntriesPerRow,Indices,srcValues,y,x,numMyRows)
 #endif
     for(int i = 0; i < numMyRows; i++) {
       int     NumEntries = NumEntriesPerRow[i];
@@ -3395,7 +3395,7 @@ if (StorageOptimized() && Graph().StorageOptimized()) {
 
     // Do actual computation
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(x,y)
+#pragma omp parallel for default(none) shared(x,y,numMyRows)
 #endif
     for(int i = 0; i < numMyRows; i++) {
       int     NumEntries = NumMyEntries(i);
@@ -3517,7 +3517,7 @@ void Epetra_CrsMatrix::GeneralMM(double ** X, int LDX, double ** Y, int LDY, int
     double ** const yp = Y;
     const int numMyRows = NumMyRows_;
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(IndexOffset,Indices,values,NumVectors)
+#pragma omp parallel for default(none) shared(IndexOffset,Indices,values,NumVectors,numMyRows,xp,yp)
 #endif
     for (int i=0; i < numMyRows; i++) {
       int prevOffset = IndexOffset[i];
@@ -3546,7 +3546,7 @@ void Epetra_CrsMatrix::GeneralMM(double ** X, int LDX, double ** Y, int LDY, int
     const int numMyRows = NumMyRows_;
 
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(NumEntriesPerRow,Indices,srcValues,NumVectors)
+#pragma omp parallel for default(none) shared(NumEntriesPerRow,Indices,srcValues,NumVectors,numMyRows,xp,yp)
 #endif
     for (int i=0; i < numMyRows; i++) {
       int      NumEntries = NumEntriesPerRow[i];
@@ -3567,7 +3567,7 @@ void Epetra_CrsMatrix::GeneralMM(double ** X, int LDX, double ** Y, int LDY, int
     double ** const yp = Y;
     const int numMyRows = NumMyRows_;
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(NumVectors)
+#pragma omp parallel for default(none) shared(NumVectors,numMyRows,xp,yp)
 #endif
     for (int i=0; i < numMyRows; i++) {
       int     NumEntries = NumMyEntries(i);

--- a/packages/epetra/src/Epetra_IntMultiVector.cpp
+++ b/packages/epetra/src/Epetra_IntMultiVector.cpp
@@ -291,7 +291,7 @@ int Epetra_IntMultiVector::DoCopy(void)
       int * from = Pointers_[i];
       int * to = Values_+i*Stride_;
       Pointers_[i] = to;
-      const int myLength = MyLength_;
+      int myLength = MyLength_;
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(to,from,myLength)
       for (int j=0; j<myLength; j++) to[j] = from[j];
@@ -569,9 +569,9 @@ int Epetra_IntMultiVector::PutScalar(int ScalarConstant) {
 
   // Fills MultiVector with the value ScalarConstant **/
 
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
   for (int i = 0; i < NumVectors_; i++) {
-    int * const to = Pointers_[i];
+    int * to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarConstant,myLength,to)
 #endif
@@ -1099,12 +1099,12 @@ int  Epetra_IntMultiVector::MinValue (int* Result) const {
 
   int ierr = 0;
 
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
   UpdateOrdinalTemp();
 
   for (int i=0; i < NumVectors_; i++)
     {
-      const int * const from = Pointers_[i];
+      const int * from = Pointers_[i];
       int MinVal = 2000000000; // 2 billion is close to largest 32 bit int
       if (myLength>0) MinVal = from[0];
 #ifdef EPETRA_HAVE_OMP
@@ -1212,12 +1212,12 @@ int  Epetra_IntMultiVector::MaxValue (int* Result) const {
 
   int ierr = 0;
 
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
   UpdateOrdinalTemp();
 
   for (int i=0; i < NumVectors_; i++)
     {
-      const int * const from = Pointers_[i];
+      const int * from = Pointers_[i];
       int MaxVal = -2000000000; // Negative 2 billion is close to smallest 32 bit int
       if (myLength>0) MaxVal = from[0];
 #ifdef EPETRA_HAVE_OMP
@@ -1364,7 +1364,7 @@ Epetra_IntMultiVector& Epetra_IntMultiVector::operator = (const Epetra_IntMultiV
 //=========================================================================
 void Epetra_IntMultiVector::Assign(const Epetra_IntMultiVector& A) {
 
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
   if (NumVectors_ != A.NumVectors())
     throw ReportError("Number of vectors incompatible in Assign.  The this MultiVector has NumVectors = " + toString(NumVectors_)
           + ".  The A MultiVector has NumVectors = " + toString(A.NumVectors()), -3);
@@ -1374,8 +1374,8 @@ void Epetra_IntMultiVector::Assign(const Epetra_IntMultiVector& A) {
 
   int ** A_Pointers = A.Pointers();
   for (int i = 0; i< NumVectors_; i++) {
-      int * const to = Pointers_[i];
-      const int * const from = A_Pointers[i];
+      int * to = Pointers_[i];
+      const int * from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(myLength,to,from)
 #endif

--- a/packages/epetra/src/Epetra_IntMultiVector.cpp
+++ b/packages/epetra/src/Epetra_IntMultiVector.cpp
@@ -293,7 +293,7 @@ int Epetra_IntMultiVector::DoCopy(void)
       Pointers_[i] = to;
       const int myLength = MyLength_;
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(to,from)
+#pragma omp parallel for default(none) shared(to,from,myLength)
       for (int j=0; j<myLength; j++) to[j] = from[j];
 #else
       memcpy(to, from, myLength*sizeof(int));
@@ -573,7 +573,7 @@ int Epetra_IntMultiVector::PutScalar(int ScalarConstant) {
   for (int i = 0; i < NumVectors_; i++) {
     int * const to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarConstant)
+#pragma omp parallel for default(none) shared(ScalarConstant,myLength,to)
 #endif
     for (int j=0; j<myLength; j++) to[j] = ScalarConstant;
   }
@@ -1108,7 +1108,7 @@ int  Epetra_IntMultiVector::MinValue (int* Result) const {
       int MinVal = 2000000000; // 2 billion is close to largest 32 bit int
       if (myLength>0) MinVal = from[0];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel default(none) shared(MinVal)
+#pragma omp parallel default(none) shared(MinVal,myLength,from)
 {
       int localMinVal = MinVal;
 #pragma omp for
@@ -1221,7 +1221,7 @@ int  Epetra_IntMultiVector::MaxValue (int* Result) const {
       int MaxVal = -2000000000; // Negative 2 billion is close to smallest 32 bit int
       if (myLength>0) MaxVal = from[0];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel default(none) shared(MaxVal)
+#pragma omp parallel default(none) shared(MaxVal,myLength,from)
 {
       int localMaxVal = MaxVal;
 #pragma omp for
@@ -1377,7 +1377,7 @@ void Epetra_IntMultiVector::Assign(const Epetra_IntMultiVector& A) {
       int * const to = Pointers_[i];
       const int * const from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none)
+#pragma omp parallel for default(none) shared(myLength,to,from)
 #endif
       for (int j=0; j<myLength; j++) to[j] = from[j];
     }

--- a/packages/epetra/src/Epetra_MultiVector.cpp
+++ b/packages/epetra/src/Epetra_MultiVector.cpp
@@ -291,7 +291,7 @@ int Epetra_MultiVector::DoCopy(void)
       double * from = Pointers_[i];
       double * to = Values_+i*Stride_;
       Pointers_[i] = to;
-      const int myLength = MyLength_;
+      int myLength = MyLength_;
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(to,from,myLength)
       for (int j=0; j<myLength; j++) to[j] = from[j];
@@ -596,9 +596,9 @@ int Epetra_MultiVector::PutScalar(double ScalarConstant) {
 
   // Fills MultiVector with the value ScalarConstant **/
 
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
   for (int i = 0; i < NumVectors_; i++) {
-    double * const to = Pointers_[i];
+    double * to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarConstant,myLength,to)
 #endif
@@ -1159,15 +1159,15 @@ int Epetra_MultiVector::Abs(const Epetra_MultiVector& A) {
 
   // this[i][j] = std::abs(A[i][j])
 
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
   if (NumVectors_ != A.NumVectors()) EPETRA_CHK_ERR(-1);
   if (myLength != A.MyLength()) EPETRA_CHK_ERR(-2);
 
   double **A_Pointers = A.Pointers();
 
   for (int i=0; i < NumVectors_; i++) {
-    double * const to = Pointers_[i];
-    const double * const from = A_Pointers[i];
+    double * to = Pointers_[i];
+    const double * from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(myLength,to,from)
 #endif
@@ -1185,15 +1185,15 @@ int Epetra_MultiVector::Reciprocal(const Epetra_MultiVector& A) {
 #ifndef EPETRA_HAVE_OMP
   int localierr = 0;
 #endif
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
   if (NumVectors_ != A.NumVectors()) EPETRA_CHK_ERR(-1);
   if (myLength != A.MyLength()) EPETRA_CHK_ERR(-2);
 
   double **A_Pointers = A.Pointers();
 
   for (int i=0; i < NumVectors_; i++) {
-    double * const to = Pointers_[i];
-    const double * const from = A_Pointers[i];
+    double * to = Pointers_[i];
+    const double * from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel default(none) shared(ierr,myLength,to,from)
 {
@@ -1231,10 +1231,10 @@ int Epetra_MultiVector::Reciprocal(const Epetra_MultiVector& A) {
     // scales a MultiVector in place by a scalar
 
 
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
 #ifdef EPETRA_HAVE_OMP
     for (int i = 0; i < NumVectors_; i++) {
-      double * const to = Pointers_[i];
+      double * to = Pointers_[i];
 #pragma omp parallel for default(none) shared(ScalarValue,myLength,to)
       for (int j = 0; j < myLength; j++) to[j] = ScalarValue * to[j];
     }
@@ -1253,15 +1253,15 @@ int Epetra_MultiVector::Reciprocal(const Epetra_MultiVector& A) {
     // scales a MultiVector by a scalar and put in the this:
     // this = ScalarA * A
 
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
   if (NumVectors_ != A.NumVectors()) EPETRA_CHK_ERR(-1);
   if (myLength != A.MyLength()) EPETRA_CHK_ERR(-2);
 
     double **A_Pointers = (double**)A.Pointers();
 
     for (int i = 0; i < NumVectors_; i++) {
-      double * const to = Pointers_[i];
-      const double * const from = A_Pointers[i];
+      double * to = Pointers_[i];
+      const double * from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarA,myLength,to,from)
 #endif
@@ -1278,7 +1278,7 @@ int Epetra_MultiVector::Reciprocal(const Epetra_MultiVector& A) {
 
     // linear combination of two MultiVectors: this = ScalarThis * this + ScalarA * A
 
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
   if (NumVectors_ != A.NumVectors()) EPETRA_CHK_ERR(-1);
   if (myLength != A.MyLength()) EPETRA_CHK_ERR(-2);
 
@@ -1287,8 +1287,8 @@ int Epetra_MultiVector::Reciprocal(const Epetra_MultiVector& A) {
     if (ScalarThis==0.0)
       {
   for (int i = 0; i < NumVectors_; i++) {
-          double * const to = Pointers_[i];
-          const double * const from = A_Pointers[i];
+          double * to = Pointers_[i];
+          const double * from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarA,myLength,to,from)
 #endif
@@ -1299,8 +1299,8 @@ int Epetra_MultiVector::Reciprocal(const Epetra_MultiVector& A) {
     else if (ScalarThis==1.0)
       {
   for (int i = 0; i < NumVectors_; i++) {
-          double * const to = Pointers_[i];
-          const double * const from = A_Pointers[i];
+          double * to = Pointers_[i];
+          const double * from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarA,to,from,myLength)
 #endif
@@ -1311,8 +1311,8 @@ int Epetra_MultiVector::Reciprocal(const Epetra_MultiVector& A) {
     else if (ScalarA==1.0)
       {
   for (int i = 0; i < NumVectors_; i++) {
-          double * const to = Pointers_[i];
-          const double * const from = A_Pointers[i];
+          double * to = Pointers_[i];
+          const double * from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarThis,myLength,to,from)
 #endif
@@ -1323,8 +1323,8 @@ int Epetra_MultiVector::Reciprocal(const Epetra_MultiVector& A) {
     else
       {
   for (int i = 0; i < NumVectors_; i++) {
-          double * const to = Pointers_[i];
-          const double * const from = A_Pointers[i];
+          double * to = Pointers_[i];
+          const double * from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarA,ScalarThis,to,from,myLength)
 #endif
@@ -1354,7 +1354,7 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
     return(0);
   }
 
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
   if (NumVectors_ != A.NumVectors() || NumVectors_ != B.NumVectors()) EPETRA_CHK_ERR(-1);
   if (myLength != A.MyLength() || myLength != B.MyLength()) EPETRA_CHK_ERR(-2);
 
@@ -1366,9 +1366,9 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
   if (ScalarA==1.0)
     {
       for (int i = 0; i < NumVectors_; i++) {
-              double * const to = Pointers_[i];
-              const double * const fromA = A_Pointers[i];
-              const double * const fromB = B_Pointers[i];
+              double * to = Pointers_[i];
+              const double * fromA = A_Pointers[i];
+              const double * fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarB,myLength,to,fromA,fromB)
 #endif
@@ -1380,9 +1380,9 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
   else if (ScalarB==1.0)
     {
       for (int i = 0; i < NumVectors_; i++) {
-              double * const to = Pointers_[i];
-              const double * const fromA = A_Pointers[i];
-              const double * const fromB = B_Pointers[i];
+              double * to = Pointers_[i];
+              const double * fromA = A_Pointers[i];
+              const double * fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarA,myLength,to,fromA,fromB)
 #endif
@@ -1394,9 +1394,9 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
   else
     {
       for (int i = 0; i < NumVectors_; i++) {
-              double * const to = Pointers_[i];
-              const double * const fromA = A_Pointers[i];
-              const double * const fromB = B_Pointers[i];
+              double * to = Pointers_[i];
+              const double * fromA = A_Pointers[i];
+              const double * fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarA,ScalarB,myLength,to,fromA,fromB)
 #endif
@@ -1411,9 +1411,9 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
   if (ScalarA==1.0)
     {
       for (int i = 0; i < NumVectors_; i++) {
-              double * const to = Pointers_[i];
-              const double * const fromA = A_Pointers[i];
-              const double * const fromB = B_Pointers[i];
+              double * to = Pointers_[i];
+              const double * fromA = A_Pointers[i];
+              const double * fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarB,myLength,to,fromA,fromB)
 #endif
@@ -1425,9 +1425,9 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
   else if (ScalarB==1.0)
     {
       for (int i = 0; i < NumVectors_; i++) {
-              double * const to = Pointers_[i];
-              const double * const fromA = A_Pointers[i];
-              const double * const fromB = B_Pointers[i];
+              double * to = Pointers_[i];
+              const double * fromA = A_Pointers[i];
+              const double * fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarA,myLength,to,fromA,fromB)
 #endif
@@ -1439,9 +1439,9 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
   else
     {
       for (int i = 0; i < NumVectors_; i++) {
-              double * const to = Pointers_[i];
-              const double * const fromA = A_Pointers[i];
-              const double * const fromB = B_Pointers[i];
+              double * to = Pointers_[i];
+              const double * fromA = A_Pointers[i];
+              const double * fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarA,ScalarB,myLength,to,fromA,fromB)
 #endif
@@ -1456,9 +1456,9 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
   if (ScalarA==1.0)
     {
       for (int i = 0; i < NumVectors_; i++) {
-              double * const to = Pointers_[i];
-              const double * const fromA = A_Pointers[i];
-              const double * const fromB = B_Pointers[i];
+              double * to = Pointers_[i];
+              const double * fromA = A_Pointers[i];
+              const double * fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarA,ScalarB,ScalarThis,myLength,to,fromA,fromB)
 #endif
@@ -1471,9 +1471,9 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
   else if (ScalarB==1.0)
     {
       for (int i = 0; i < NumVectors_; i++) {
-              double * const to = Pointers_[i];
-              const double * const fromA = A_Pointers[i];
-              const double * const fromB = B_Pointers[i];
+              double * to = Pointers_[i];
+              const double * fromA = A_Pointers[i];
+              const double * fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarA,ScalarThis,myLength,to,fromA,fromB)
 #endif
@@ -1486,9 +1486,9 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
   else
     {
       for (int i = 0; i < NumVectors_; i++) {
-              double * const to = Pointers_[i];
-              const double * const fromA = A_Pointers[i];
-              const double * const fromB = B_Pointers[i];
+              double * to = Pointers_[i];
+              const double * fromA = A_Pointers[i];
+              const double * fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarA,ScalarB,ScalarThis,myLength,to,fromA,fromB)
 #endif
@@ -1511,12 +1511,12 @@ int  Epetra_MultiVector::Norm1 (double* Result) const {
 
   if (!Map().UniqueGIDs()) {EPETRA_CHK_ERR(-1);}
 
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
   UpdateDoubleTemp();
 #ifdef EPETRA_HAVE_OMP_NONASSOCIATIVE
   for (int i=0; i < NumVectors_; i++)
     {
-      const double * const from = Pointers_[i];
+      const double * from = Pointers_[i];
       double asum = 0.0;
 #pragma omp parallel default(none) shared(asum,myLength)
 {
@@ -1582,14 +1582,14 @@ int  Epetra_MultiVector::NormInf (double* Result) const {
   // Inf-norm of each vector in MultiVector
 
 
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
   UpdateDoubleTemp();
 
   for (int i=0; i < NumVectors_; i++)
     {
       DoubleTemp_[i] = 0.0;
       double normval = 0.0;
-      const double * const from = Pointers_[i];
+      const double * from = Pointers_[i];
       if (myLength>0) normval = from[0];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel default(none) shared(normval,myLength,from)
@@ -1627,7 +1627,7 @@ int  Epetra_MultiVector::NormWeighted (const Epetra_MultiVector& Weights, double
 
   // If only one vector in Weights, we assume it will be used as the weights for all vectors
 
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
   bool OneW = false;
   if (Weights.NumVectors()==1) OneW = true;
   else if (NumVectors_ != Weights.NumVectors()) EPETRA_CHK_ERR(-1);
@@ -1644,7 +1644,7 @@ int  Epetra_MultiVector::NormWeighted (const Epetra_MultiVector& Weights, double
     {
       if (!OneW) W = W_Pointers[i]; // If Weights has the same number of vectors as this, use each weight vector
       double sum = 0.0;
-      const double * const from = Pointers_[i];
+      const double * from = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP_NONASSOCIATIVE
 #pragma omp parallel for reduction (+:sum) default(none) shared(W)
 #endif
@@ -1677,12 +1677,12 @@ int  Epetra_MultiVector::MinValue (double* Result) const {
 
   int ierr = 0;
 
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
   UpdateDoubleTemp();
 
   for (int i=0; i < NumVectors_; i++)
     {
-      const double * const from = Pointers_[i];
+      const double * from = Pointers_[i];
       double MinVal = Epetra_MaxDouble;
       if (myLength>0) MinVal = from[0];
 #ifdef EPETRA_HAVE_OMP
@@ -1790,12 +1790,12 @@ int  Epetra_MultiVector::MaxValue (double* Result) const {
 
   int ierr = 0;
 
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
   UpdateDoubleTemp();
 
   for (int i=0; i < NumVectors_; i++)
     {
-      const double * const from = Pointers_[i];
+      const double * from = Pointers_[i];
       double MaxVal = -Epetra_MaxDouble;
       if (myLength>0) MaxVal = from[0];
 #ifdef EPETRA_HAVE_OMP
@@ -2095,7 +2095,7 @@ int Epetra_MultiVector::Multiply(double ScalarAB, const Epetra_MultiVector& A, c
     EPETRA_CHK_ERR(Scale(ScalarThis));
     return(0);
   }
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
 
   if (A.NumVectors() != 1 && A.NumVectors() != B.NumVectors()) EPETRA_CHK_ERR(-1); // A must have one column or be the same as B.
   if (NumVectors_ != B.NumVectors()) EPETRA_CHK_ERR(-2);
@@ -2111,9 +2111,9 @@ int Epetra_MultiVector::Multiply(double ScalarAB, const Epetra_MultiVector& A, c
       if (ScalarAB==1.0)
   {
     for (int i = 0; i < NumVectors_; i++) {
-      const double * const Aptr = A_Pointers[i*IncA];
-            const double * const Bptr = B_Pointers[i];
-            double * const to = Pointers_[i];
+      const double * Aptr = A_Pointers[i*IncA];
+      const double * Bptr = B_Pointers[i];
+      double * to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(myLength,to,Aptr,Bptr)
 #endif
@@ -2126,9 +2126,9 @@ int Epetra_MultiVector::Multiply(double ScalarAB, const Epetra_MultiVector& A, c
       else
   {
     for (int i = 0; i < NumVectors_; i++) {
-      const double * const Aptr = A_Pointers[i*IncA];
-            const double * const Bptr = B_Pointers[i];
-            double * const to = Pointers_[i];
+      const double * Aptr = A_Pointers[i*IncA];
+      const double * Bptr = B_Pointers[i];
+      double * to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarAB,myLength,to,Aptr,Bptr)
 #endif
@@ -2143,9 +2143,9 @@ int Epetra_MultiVector::Multiply(double ScalarAB, const Epetra_MultiVector& A, c
       if (ScalarAB==1.0)
   {
     for (int i = 0; i < NumVectors_; i++) {
-      const double * const Aptr = A_Pointers[i*IncA];
-            const double * const Bptr = B_Pointers[i];
-            double * const to = Pointers_[i];
+      const double * Aptr = A_Pointers[i*IncA];
+      const double * Bptr = B_Pointers[i];
+      double * to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(myLength,to,Aptr,Bptr)
 #endif
@@ -2157,9 +2157,9 @@ int Epetra_MultiVector::Multiply(double ScalarAB, const Epetra_MultiVector& A, c
   }
       else {
     for (int i = 0; i < NumVectors_; i++) {
-      const double * const Aptr = A_Pointers[i*IncA];
-            const double * const Bptr = B_Pointers[i];
-            double * const to = Pointers_[i];
+      const double * Aptr = A_Pointers[i*IncA];
+      const double * Bptr = B_Pointers[i];
+      double * to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarAB,myLength,to,Aptr,Bptr)
 #endif
@@ -2174,9 +2174,9 @@ int Epetra_MultiVector::Multiply(double ScalarAB, const Epetra_MultiVector& A, c
       if (ScalarAB==1.0)
   {
     for (int i = 0; i < NumVectors_; i++) {
-      const double * const Aptr = A_Pointers[i*IncA];
-            const double * const Bptr = B_Pointers[i];
-            double * const to = Pointers_[i];
+      const double * Aptr = A_Pointers[i*IncA];
+      const double * Bptr = B_Pointers[i];
+      double * to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarThis,myLength,to,Aptr,Bptr)
 #endif
@@ -2190,9 +2190,9 @@ int Epetra_MultiVector::Multiply(double ScalarAB, const Epetra_MultiVector& A, c
       else
   {
     for (int i = 0; i < NumVectors_; i++) {
-      const double * const Aptr = A_Pointers[i*IncA];
-            const double * const Bptr = B_Pointers[i];
-            double * const to = Pointers_[i];
+      const double * Aptr = A_Pointers[i*IncA];
+      const double * Bptr = B_Pointers[i];
+      double * to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarThis,ScalarAB,myLength,to,Aptr,Bptr)
 #endif
@@ -2218,7 +2218,7 @@ int Epetra_MultiVector::ReciprocalMultiply(double ScalarAB, const Epetra_MultiVe
     EPETRA_CHK_ERR(Scale(ScalarThis));
     return(0);
   }
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
 
   if (A.NumVectors() != 1 && A.NumVectors() != B.NumVectors()) EPETRA_CHK_ERR(-1); // A must have one column or be the same as B.
   if (NumVectors_ != B.NumVectors()) EPETRA_CHK_ERR(-2);
@@ -2234,9 +2234,9 @@ int Epetra_MultiVector::ReciprocalMultiply(double ScalarAB, const Epetra_MultiVe
       if (ScalarAB==1.0)
   {
     for (int i = 0; i < NumVectors_; i++) {
-      const double * const Aptr = A_Pointers[i*IncA];
-            const double * const Bptr = B_Pointers[i];
-            double * const to = Pointers_[i];
+      const double * Aptr = A_Pointers[i*IncA];
+      const double * Bptr = B_Pointers[i];
+      double * to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(myLength,to,Aptr,Bptr)
 #endif
@@ -2249,9 +2249,9 @@ int Epetra_MultiVector::ReciprocalMultiply(double ScalarAB, const Epetra_MultiVe
       else
   {
     for (int i = 0; i < NumVectors_; i++) {
-      const double * const Aptr = A_Pointers[i*IncA];
-            const double * const Bptr = B_Pointers[i];
-            double * const to = Pointers_[i];
+      const double * Aptr = A_Pointers[i*IncA];
+      const double * Bptr = B_Pointers[i];
+      double * to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarAB,myLength,to,Aptr,Bptr)
 #endif
@@ -2266,9 +2266,9 @@ int Epetra_MultiVector::ReciprocalMultiply(double ScalarAB, const Epetra_MultiVe
       if (ScalarAB==1.0)
   {
     for (int i = 0; i < NumVectors_; i++) {
-      const double * const Aptr = A_Pointers[i*IncA];
-            const double * const Bptr = B_Pointers[i];
-            double * const to = Pointers_[i];
+      const double * Aptr = A_Pointers[i*IncA];
+      const double * Bptr = B_Pointers[i];
+      double * to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(myLength,to,Aptr,Bptr)
 #endif
@@ -2281,9 +2281,9 @@ int Epetra_MultiVector::ReciprocalMultiply(double ScalarAB, const Epetra_MultiVe
       else
   {
     for (int i = 0; i < NumVectors_; i++) {
-      const double * const Aptr = A_Pointers[i*IncA];
-            const double * const Bptr = B_Pointers[i];
-            double * const to = Pointers_[i];
+      const double * Aptr = A_Pointers[i*IncA];
+      const double * Bptr = B_Pointers[i];
+      double * to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarAB,myLength,to,Aptr,Bptr)
 #endif
@@ -2298,9 +2298,9 @@ int Epetra_MultiVector::ReciprocalMultiply(double ScalarAB, const Epetra_MultiVe
       if (ScalarAB==1.0)
   {
     for (int i = 0; i < NumVectors_; i++) {
-      const double * const Aptr = A_Pointers[i*IncA];
-            const double * const Bptr = B_Pointers[i];
-            double * const to = Pointers_[i];
+      const double * Aptr = A_Pointers[i*IncA];
+      const double * Bptr = B_Pointers[i];
+      double * to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarThis,myLength,to,Aptr,Bptr)
 #endif
@@ -2313,9 +2313,9 @@ int Epetra_MultiVector::ReciprocalMultiply(double ScalarAB, const Epetra_MultiVe
   }
       else {
     for (int i = 0; i < NumVectors_; i++) {
-      const double * const Aptr = A_Pointers[i*IncA];
-            const double * const Bptr = B_Pointers[i];
-            double * const to = Pointers_[i];
+      const double * Aptr = A_Pointers[i*IncA];
+      const double * Bptr = B_Pointers[i];
+      double * to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(ScalarAB,ScalarThis,myLength,to,Aptr,Bptr)
 #endif
@@ -2376,7 +2376,7 @@ Epetra_MultiVector& Epetra_MultiVector::operator = (const Epetra_MultiVector& So
 //=========================================================================
 void Epetra_MultiVector::Assign(const Epetra_MultiVector& A) {
 
-  const int myLength = MyLength_;
+  int myLength = MyLength_;
   if (NumVectors_ != A.NumVectors())
     throw ReportError("Number of vectors incompatible in Assign.  The this MultiVector has NumVectors = " + toString(NumVectors_)
           + ".  The A MultiVector has NumVectors = " + toString(A.NumVectors()), -3);
@@ -2386,8 +2386,8 @@ void Epetra_MultiVector::Assign(const Epetra_MultiVector& A) {
 
   double ** A_Pointers = A.Pointers();
   for (int i = 0; i< NumVectors_; i++) {
-      double * const to = Pointers_[i];
-      const double * const from = A_Pointers[i];
+      double * to = Pointers_[i];
+      const double * from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
 #pragma omp parallel for default(none) shared(myLength,to,from)
 #endif

--- a/packages/epetra/src/Epetra_MultiVector.cpp
+++ b/packages/epetra/src/Epetra_MultiVector.cpp
@@ -293,7 +293,7 @@ int Epetra_MultiVector::DoCopy(void)
       Pointers_[i] = to;
       const int myLength = MyLength_;
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(to,from)
+#pragma omp parallel for default(none) shared(to,from,myLength)
       for (int j=0; j<myLength; j++) to[j] = from[j];
 #else
       memcpy(to, from, myLength*sizeof(double));
@@ -600,7 +600,7 @@ int Epetra_MultiVector::PutScalar(double ScalarConstant) {
   for (int i = 0; i < NumVectors_; i++) {
     double * const to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarConstant)
+#pragma omp parallel for default(none) shared(ScalarConstant,myLength,to)
 #endif
     for (int j=0; j<myLength; j++) to[j] = ScalarConstant;
   }
@@ -1169,7 +1169,7 @@ int Epetra_MultiVector::Abs(const Epetra_MultiVector& A) {
     double * const to = Pointers_[i];
     const double * const from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none)
+#pragma omp parallel for default(none) shared(myLength,to,from)
 #endif
     for (int j=0; j < myLength; j++) to[j] = std::abs(from[j]);
   }
@@ -1195,7 +1195,7 @@ int Epetra_MultiVector::Reciprocal(const Epetra_MultiVector& A) {
     double * const to = Pointers_[i];
     const double * const from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel default(none) shared(ierr)
+#pragma omp parallel default(none) shared(ierr,myLength,to,from)
 {
     int localierr = 0;
 #pragma omp for
@@ -1235,7 +1235,7 @@ int Epetra_MultiVector::Reciprocal(const Epetra_MultiVector& A) {
 #ifdef EPETRA_HAVE_OMP
     for (int i = 0; i < NumVectors_; i++) {
       double * const to = Pointers_[i];
-#pragma omp parallel for default(none) shared(ScalarValue)
+#pragma omp parallel for default(none) shared(ScalarValue,myLength,to)
       for (int j = 0; j < myLength; j++) to[j] = ScalarValue * to[j];
     }
 #else
@@ -1263,7 +1263,7 @@ int Epetra_MultiVector::Reciprocal(const Epetra_MultiVector& A) {
       double * const to = Pointers_[i];
       const double * const from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarA)
+#pragma omp parallel for default(none) shared(ScalarA,myLength,to,from)
 #endif
       for (int j = 0; j < myLength; j++) to[j] = ScalarA * from[j];
     }
@@ -1290,7 +1290,7 @@ int Epetra_MultiVector::Reciprocal(const Epetra_MultiVector& A) {
           double * const to = Pointers_[i];
           const double * const from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarA)
+#pragma omp parallel for default(none) shared(ScalarA,myLength,to,from)
 #endif
     for (int j = 0; j < myLength; j++) to[j] = ScalarA * from[j];
         }
@@ -1302,7 +1302,7 @@ int Epetra_MultiVector::Reciprocal(const Epetra_MultiVector& A) {
           double * const to = Pointers_[i];
           const double * const from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarA)
+#pragma omp parallel for default(none) shared(ScalarA,to,from,myLength)
 #endif
     for (int j = 0; j < myLength; j++) to[j] = to[j] + ScalarA * from[j];
         }
@@ -1314,7 +1314,7 @@ int Epetra_MultiVector::Reciprocal(const Epetra_MultiVector& A) {
           double * const to = Pointers_[i];
           const double * const from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarThis)
+#pragma omp parallel for default(none) shared(ScalarThis,myLength,to,from)
 #endif
     for (int j = 0; j < myLength; j++) to[j] = ScalarThis * to[j] + from[j];
         }
@@ -1326,7 +1326,7 @@ int Epetra_MultiVector::Reciprocal(const Epetra_MultiVector& A) {
           double * const to = Pointers_[i];
           const double * const from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarA,ScalarThis)
+#pragma omp parallel for default(none) shared(ScalarA,ScalarThis,to,from,myLength)
 #endif
     for (int j = 0; j < myLength; j++) to[j] = ScalarThis * to[j] +
                               ScalarA *  from[j];
@@ -1370,7 +1370,7 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
               const double * const fromA = A_Pointers[i];
               const double * const fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarB)
+#pragma omp parallel for default(none) shared(ScalarB,myLength,to,fromA,fromB)
 #endif
         for (int j = 0; j < myLength; j++) to[j] =           fromA[j] +
                             ScalarB * fromB[j];
@@ -1384,7 +1384,7 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
               const double * const fromA = A_Pointers[i];
               const double * const fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarA)
+#pragma omp parallel for default(none) shared(ScalarA,myLength,to,fromA,fromB)
 #endif
         for (int j = 0; j < myLength; j++) to[j] = ScalarA * fromA[j] +
                                       fromB[j];
@@ -1398,7 +1398,7 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
               const double * const fromA = A_Pointers[i];
               const double * const fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarA,ScalarB)
+#pragma omp parallel for default(none) shared(ScalarA,ScalarB,myLength,to,fromA,fromB)
 #endif
         for (int j = 0; j < myLength; j++) to[j] = ScalarA * fromA[j] +
                             ScalarB * fromB[j];
@@ -1415,7 +1415,7 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
               const double * const fromA = A_Pointers[i];
               const double * const fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarB)
+#pragma omp parallel for default(none) shared(ScalarB,myLength,to,fromA,fromB)
 #endif
         for (int j = 0; j < myLength; j++) to[j] +=           fromA[j] +
                              ScalarB * fromB[j];
@@ -1429,7 +1429,7 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
               const double * const fromA = A_Pointers[i];
               const double * const fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarA)
+#pragma omp parallel for default(none) shared(ScalarA,myLength,to,fromA,fromB)
 #endif
         for (int j = 0; j < myLength; j++) to[j] += ScalarA * fromA[j] +
                                        fromB[j];
@@ -1443,7 +1443,7 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
               const double * const fromA = A_Pointers[i];
               const double * const fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarA,ScalarB)
+#pragma omp parallel for default(none) shared(ScalarA,ScalarB,myLength,to,fromA,fromB)
 #endif
         for (int j = 0; j < myLength; j++) to[j] += ScalarA * fromA[j] +
                              ScalarB * fromB[j];
@@ -1460,7 +1460,7 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
               const double * const fromA = A_Pointers[i];
               const double * const fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarA,ScalarB,ScalarThis)
+#pragma omp parallel for default(none) shared(ScalarA,ScalarB,ScalarThis,myLength,to,fromA,fromB)
 #endif
         for (int j = 0; j < myLength; j++) to[j] =  ScalarThis *    to[j] +
                                        fromA[j] +
@@ -1475,7 +1475,7 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
               const double * const fromA = A_Pointers[i];
               const double * const fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarA,ScalarThis)
+#pragma omp parallel for default(none) shared(ScalarA,ScalarThis,myLength,to,fromA,fromB)
 #endif
         for (int j = 0; j < myLength; j++) to[j] =  ScalarThis *    to[j] +
                              ScalarA * fromA[j] +
@@ -1490,7 +1490,7 @@ int Epetra_MultiVector::Update(double ScalarA, const Epetra_MultiVector& A,
               const double * const fromA = A_Pointers[i];
               const double * const fromB = B_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarA,ScalarB,ScalarThis)
+#pragma omp parallel for default(none) shared(ScalarA,ScalarB,ScalarThis,myLength,to,fromA,fromB)
 #endif
         for (int j = 0; j < myLength; j++) to[j] =  ScalarThis *    to[j] +
                              ScalarA * fromA[j] +
@@ -1518,7 +1518,7 @@ int  Epetra_MultiVector::Norm1 (double* Result) const {
     {
       const double * const from = Pointers_[i];
       double asum = 0.0;
-#pragma omp parallel default(none) shared(asum)
+#pragma omp parallel default(none) shared(asum,myLength)
 {
       double localasum = 0.0;
 #pragma omp for
@@ -1592,7 +1592,7 @@ int  Epetra_MultiVector::NormInf (double* Result) const {
       const double * const from = Pointers_[i];
       if (myLength>0) normval = from[0];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel default(none) shared(normval)
+#pragma omp parallel default(none) shared(normval,myLength,from)
 {
       double localnormval = 0.0;
 #pragma omp for
@@ -1686,7 +1686,7 @@ int  Epetra_MultiVector::MinValue (double* Result) const {
       double MinVal = Epetra_MaxDouble;
       if (myLength>0) MinVal = from[0];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel default(none) shared(MinVal)
+#pragma omp parallel default(none) shared(MinVal,from,myLength)
 {
       double localMinVal = MinVal;
 #pragma omp for
@@ -1799,7 +1799,7 @@ int  Epetra_MultiVector::MaxValue (double* Result) const {
       double MaxVal = -Epetra_MaxDouble;
       if (myLength>0) MaxVal = from[0];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel default(none) shared(MaxVal)
+#pragma omp parallel default(none) shared(MaxVal,myLength,from)
 {
       double localMaxVal = MaxVal;
 #pragma omp for
@@ -2115,7 +2115,7 @@ int Epetra_MultiVector::Multiply(double ScalarAB, const Epetra_MultiVector& A, c
             const double * const Bptr = B_Pointers[i];
             double * const to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none)
+#pragma omp parallel for default(none) shared(myLength,to,Aptr,Bptr)
 #endif
       for (int j = 0; j < myLength; j++) {
               to[j] =  Aptr[j] * Bptr[j];
@@ -2130,7 +2130,7 @@ int Epetra_MultiVector::Multiply(double ScalarAB, const Epetra_MultiVector& A, c
             const double * const Bptr = B_Pointers[i];
             double * const to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarAB)
+#pragma omp parallel for default(none) shared(ScalarAB,myLength,to,Aptr,Bptr)
 #endif
       for (int j = 0; j < myLength; j++) {
               to[j] = ScalarAB * Aptr[j] * Bptr[j];
@@ -2147,7 +2147,7 @@ int Epetra_MultiVector::Multiply(double ScalarAB, const Epetra_MultiVector& A, c
             const double * const Bptr = B_Pointers[i];
             double * const to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none)
+#pragma omp parallel for default(none) shared(myLength,to,Aptr,Bptr)
 #endif
       for (int j = 0; j < myLength; j++) {
               to[j] +=  Aptr[j] * Bptr[j];
@@ -2161,7 +2161,7 @@ int Epetra_MultiVector::Multiply(double ScalarAB, const Epetra_MultiVector& A, c
             const double * const Bptr = B_Pointers[i];
             double * const to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarAB)
+#pragma omp parallel for default(none) shared(ScalarAB,myLength,to,Aptr,Bptr)
 #endif
       for (int j = 0; j < myLength; j++) {
               to[j] += ScalarAB * Aptr[j] * Bptr[j];
@@ -2178,7 +2178,7 @@ int Epetra_MultiVector::Multiply(double ScalarAB, const Epetra_MultiVector& A, c
             const double * const Bptr = B_Pointers[i];
             double * const to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarThis)
+#pragma omp parallel for default(none) shared(ScalarThis,myLength,to,Aptr,Bptr)
 #endif
       for (int j = 0; j < myLength; j++) {
               to[j] =  ScalarThis * to[j] +
@@ -2194,7 +2194,7 @@ int Epetra_MultiVector::Multiply(double ScalarAB, const Epetra_MultiVector& A, c
             const double * const Bptr = B_Pointers[i];
             double * const to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarThis,ScalarAB)
+#pragma omp parallel for default(none) shared(ScalarThis,ScalarAB,myLength,to,Aptr,Bptr)
 #endif
       for (int j = 0; j < myLength; j++) {
               to[j] = ScalarThis * to[j] +
@@ -2238,7 +2238,7 @@ int Epetra_MultiVector::ReciprocalMultiply(double ScalarAB, const Epetra_MultiVe
             const double * const Bptr = B_Pointers[i];
             double * const to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none)
+#pragma omp parallel for default(none) shared(myLength,to,Aptr,Bptr)
 #endif
       for (int j = 0; j < myLength; j++) {
               to[j] = Bptr[j] / Aptr[j];
@@ -2253,7 +2253,7 @@ int Epetra_MultiVector::ReciprocalMultiply(double ScalarAB, const Epetra_MultiVe
             const double * const Bptr = B_Pointers[i];
             double * const to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarAB)
+#pragma omp parallel for default(none) shared(ScalarAB,myLength,to,Aptr,Bptr)
 #endif
       for (int j = 0; j < myLength; j++) {
               to[j] = ScalarAB * Bptr[j] / Aptr[j];
@@ -2270,7 +2270,7 @@ int Epetra_MultiVector::ReciprocalMultiply(double ScalarAB, const Epetra_MultiVe
             const double * const Bptr = B_Pointers[i];
             double * const to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none)
+#pragma omp parallel for default(none) shared(myLength,to,Aptr,Bptr)
 #endif
       for (int j = 0; j < myLength; j++) {
               to[j] +=  Bptr[j] / Aptr[j];
@@ -2285,7 +2285,7 @@ int Epetra_MultiVector::ReciprocalMultiply(double ScalarAB, const Epetra_MultiVe
             const double * const Bptr = B_Pointers[i];
             double * const to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarAB)
+#pragma omp parallel for default(none) shared(ScalarAB,myLength,to,Aptr,Bptr)
 #endif
       for (int j = 0; j < myLength; j++) {
               to[j] += ScalarAB * Bptr[j] / Aptr[j];
@@ -2302,7 +2302,7 @@ int Epetra_MultiVector::ReciprocalMultiply(double ScalarAB, const Epetra_MultiVe
             const double * const Bptr = B_Pointers[i];
             double * const to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarThis)
+#pragma omp parallel for default(none) shared(ScalarThis,myLength,to,Aptr,Bptr)
 #endif
       for (int j = 0; j < myLength; j++) {
               to[j] =  ScalarThis * to[j] +
@@ -2317,7 +2317,7 @@ int Epetra_MultiVector::ReciprocalMultiply(double ScalarAB, const Epetra_MultiVe
             const double * const Bptr = B_Pointers[i];
             double * const to = Pointers_[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none) shared(ScalarAB,ScalarThis)
+#pragma omp parallel for default(none) shared(ScalarAB,ScalarThis,myLength,to,Aptr,Bptr)
 #endif
       for (int j = 0; j < myLength; j++) {
               to[j] = ScalarThis * to[j] + ScalarAB *
@@ -2389,7 +2389,7 @@ void Epetra_MultiVector::Assign(const Epetra_MultiVector& A) {
       double * const to = Pointers_[i];
       const double * const from = A_Pointers[i];
 #ifdef EPETRA_HAVE_OMP
-#pragma omp parallel for default(none)
+#pragma omp parallel for default(none) shared(myLength,to,from)
 #endif
       for (int j=0; j<myLength; j++) to[j] = from[j];
     }

